planner/core/explain.go

```diff
@@ -392,3 +391,156 @@ func formatWindowFuncDescs(buffer *bytes.Buffer, descs []*aggregation.WindowFunc
+	}
+	if len(p.OtherConditions) > 0 {
+		fmt.Fprintf(buffer, ", other cond:%s",
+			expression.SortedExplainExpressionList(p.OtherConditions))
+	}
+	return buffer.String()
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalAggregation) ExplainInfo() string {
+	buffer := bytes.NewBufferString("")
+	if len(p.GroupByItems) > 0 {
+		fmt.Fprintf(buffer, "group by:%s, ",
+			expression.SortedExplainExpressionList(p.GroupByItems))
+	}
+	if len(p.AggFuncs) > 0 {
+		buffer.WriteString("funcs:")
+		for i, agg := range p.AggFuncs {
+			buffer.WriteString(aggregation.ExplainAggFunc(agg))
+			if i+1 < len(p.AggFuncs) {
+				buffer.WriteString(", ")
+			}
+		}
+	}
+	return buffer.String()
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalProjection) ExplainInfo() string {
+	return string(expression.ExplainExpressionList(p.Exprs))
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalSelection) ExplainInfo() string {
+	return string(expression.SortedExplainExpressionList(p.Conditions))
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalApply) ExplainInfo() string {
+	buffer := bytes.NewBufferString("")
+	fmt.Fprintf(buffer, "%s", p.JoinType.String())
+	if len(p.EqualConditions) > 0 {
+		fmt.Fprintf(buffer, ", equal:%v", p.EqualConditions)
+	}
+	if len(p.LeftConditions) > 0 {
+		fmt.Fprintf(buffer, ", left cond:%s",
+			expression.SortedExplainExpressionList(p.LeftConditions))
+	}
+	if len(p.RightConditions) > 0 {
+		fmt.Fprintf(buffer, ", right cond:%s",
+			expression.SortedExplainExpressionList(p.RightConditions))
+	}
+	if len(p.OtherConditions) > 0 {
+		fmt.Fprintf(buffer, ", other cond:%s",
+			expression.SortedExplainExpressionList(p.OtherConditions))
+	}
+	return buffer.String()
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalTableDual) ExplainInfo() string {
+	return fmt.Sprintf("rowcount:%d", p.RowCount)
+}
+
+// ExplainInfo implements Plan interface.
+func (p *DataSource) ExplainInfo() string {
+	buffer := bytes.NewBufferString("")
+	tblName := p.tableInfo.Name.O
+	if p.TableAsName != nil && p.TableAsName.O != "" {
+		tblName = p.TableAsName.O
+	}
+	fmt.Fprintf(buffer, "table:%s", tblName)
+	if p.isPartition {
+		if pi := p.tableInfo.GetPartitionInfo(); pi != nil {
+			partitionName := pi.GetNameByID(p.physicalTableID)
+			fmt.Fprintf(buffer, ", partition:%s", partitionName)
+		}
+	}
+	if p.handleCol != nil {
+		fmt.Fprintf(buffer, ", pk col:%s", p.handleCol.ExplainInfo())
+	}
+	return buffer.String()
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalUnionScan) ExplainInfo() string {
+	buffer := bytes.NewBufferString("")
+	fmt.Fprintf(buffer, "conds:%s",
+		expression.SortedExplainExpressionList(p.conditions))
+	fmt.Fprintf(buffer, ", handle:%s", p.handleCol.ExplainInfo())
+	return buffer.String()
+}
+
+// ExplainInfo implements Plan interface.
+func (p *LogicalSort) ExplainInfo() string {
+	buffer := bytes.NewBufferString("")
+	for i, item := range p.ByItems {
```

Comment:
Extract function for `ByItems` to reuse it for `TopN`?

Reply:
All of the comments have been addressed. I also reuse the function for `Physical Plan`.